Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -288,44 +466,50 @@ void CDMInstanceOpenCDM::requestLicense(LicenseType licenseType, const AtomicStr
+            removeSession(sessionId);
+            return;
+        }
+
+        GST_DEBUG("created valid session %s", sessionId.utf8().data());
+        callback(session->message(), sessionId, session->needsIndividualization(), Succeeded);
+    };
 
     String sessionIdAsString = sessionIdByInitData(initData);
     if (!sessionIdAsString.isEmpty()) {
         auto session = lookupSession(sessionIdAsString);
-        if (session->isValid()){
+        if (session->isValid()) {
             GST_DEBUG("session %s exists and is valid, we can return now", sessionIdAsString.utf8().data());
-            callback(session->message(), sessionIdAsString, session->needsIndividualization(), Succeeded);
+            generateChallenge(session.get());
         } else {
-            GST_WARNING("existing session %s is invalid, bailing out", sessionIdAsString.utf8().data());
-            callback(WTFMove(rawInitData), sessionIdAsString, false, Failed);
+            // Created but waits on challenge.
+            session->generateChallenge(WTFMove(generateChallenge));
         }
         return;
     }
 
-    // FIXME: Why do we have this weirdness here? It looks like this is a way to reference count on the OpenCDM object.
-    media::OpenCdm openCDM(m_openCDM);
-    std::string sessionId = openCDM.CreateSession(m_mimeType, reinterpret_cast<const uint8_t*>(rawInitData->data()), rawInitData->size(), !rawCustomData->isEmpty() ? reinterpret_cast<const uint8_t*>(rawCustomData->data()) : nullptr, rawCustomData->size(), openCDMLicenseType(licenseType));
-
-    if (sessionId.empty()) {
-        GST_ERROR("could not create session id");
-        callback(WTFMove(rawInitData), { }, false, Failed);
-        return;
-    }
-
-    sessionIdAsString = String::fromUTF8(sessionId.c_str());
-    Ref<Session> newSession = Session::create(openCDM, WTFMove(rawInitData));
-
-    if (!newSession->isValid()) {
-        GST_WARNING("created invalid session %s", sessionIdAsString.utf8().data());
-        callback(WTFMove(rawInitData), sessionIdAsString, false, Failed);
+    Ref<Session> newSession = Session::create(m_openCDMAccessor, m_keySystem, m_mimeType, WTFMove(rawInitData), licenseType, WTFMove(rawCustomData));
+    String sessionId = newSession->id();
+    if (sessionId.isEmpty()) {
+        generateChallenge(newSession.ptr());
         return;
     }
+    GST_DEBUG("Created session with id %s", sessionId.utf8().data());
+    newSession->generateChallenge(WTFMove(generateChallenge));
 
-    GST_DEBUG("created valid session %s", sessionIdAsString.utf8().data());
-    callback(newSession->message(), sessionIdAsString, newSession->needsIndividualization(), Succeeded);
-
-    if (!addSession(sessionIdAsString, newSession.ptr()))
-        GST_WARNING("Failed to add session %s, the session might already exist, or the allocation failed", sessionId.c_str());
+    if (!addSession(sessionId, newSession.ptr()))
```

Comment:
I think it would make sense to `WTFMove` the session here instead of just passing the bare pointer. I guess you'd need to change the signatures of `addSession` and the hash map attribute as well. If you can't because of the hash map, turn this into `RefPtr` and keep passing that down.